2020/03/10/cooking-with-cas.html.md

```diff
@@ -0,0 +1,188 @@
+assertion comes from the CAS extension and not somewhere else. Then the `mapping` portion ([documented here](https://docs.wildfly.org/14/Admin_Guide.html#mapping))
+looks up the given user in a database to find what roles it should be assigned.
+
+The last piece of the puzzle is a CAS deployment descriptor for our application, which activates cas-extension for that application. In its simplest form, this
+is an empty file in the right place, but ours ended up a little more complex. It identifies both the CAS profile we want to use (unnecessary in this case, as
+there's only one CAS profile on the system, but it helped keep things more clear in our minds), and instructs the extension to load some other libraries into
+our application.
+
+```xml
+<?xml version="1.0" encoding="UT-8"?>
+<cas xmlns="urn:soulwing.org:cas:1.0">
+  <profile>default</profile>
+  <add-APO-dependencies/>
+</cas>
+```
+
+This configuration proved sufficient to let users log in and use the application, but as is sometimes the case with single sign-on, we needed a little more work
+to let them log out properly. Each application in the suite sets a cookie in the user's browser to identify its session. The CAS server likewise sets a cookie.
+When a user logs out of the application, that application's session cookie is destroyed, but we also need to destroy the CAS server's session cookie as well as
+the other applications' cookies. Single log-out can be complicated, and I won't go into the full setup here. One reason the CAS deployment descriptor loads
+cas-extension libraries was so we could use cas-extension to generate the proper logout URL, and redirect our users to that URL once the application has
+destroyed its own session.
+
+## Configuring Spring Authentication
+
+Another Java-based application in our suite uses Spring Security. [This document](https://docs.spring.io/spring-security/site/docs/3.0.x/reference/cas.html)
+describes the bulk of the configuration, which seemed less straightforward than for cas-extension, but follows essentially the same mechanism. Here we
+configured a Spring
+[UserDetailsService](https://docs.spring.io/spring-security/site/docs/3.2.3.RELEASE/apidocs/org/springframework/security/core/userdetails/UserDetailsService.html)
+to execute the same database query we used above to find the user's roles. I'm not fully conversant in the large stack of beans Spring uses to manage the
+process, and it took some time to get this configuration sorted out.
+
+## Enter PHP
+
+Two of these applications use PHP, which meant yet another configuration. Apereo maintains a [PHP client](https://github.com/apereo/phpCAS), which includes
+several helpful examples. Once I refreshed my memory about how to use PHP itself, I tracked down the part of the application that authenticates users, and
+replaced the existing code with calls to php CAS:
+
+```php
+    php-CAS::client(CAS_VERSION_2_0, $phpCASHost, $phpCASHost, $phpCASContext);
```

Comment:
Is the hyphen in `php-CAS` right? Seems weird when package `phpCAS` is used on the very next line.